src/types/local_date.rs

```diff
@@ -0,0 +1,271 @@
+            }
+            + day as i32
+            - 1;
+
+        if year >= 1970 {
+            total_days += (year - 1970) * 365 + Self::number_leap_years(year - 1) - LEAP_YEAR_1970
+                + number_day;
+        } else {
+            total_days -= if Self::is_leap_year(year) { 366 } else { 365 } - number_day;
+            if year < 0 {
+                total_days -= (1970 - year - 1) * 365
+                    + Self::number_leap_years(year.abs() + 1)
+                    + LEAP_YEAR_1970
+                    + 1;
+            } else {
+                total_days -=
+                    (1970 - year - 1) * 365 + LEAP_YEAR_1970 - Self::number_leap_years(year + 1);
+            }
+        }
+        total_days
+    }
+
+    /// Adds the number of days to self.
+    fn date_from_days(mut n: i64) -> (i8, i8, i32) {
+        let mut year = 1970;
+        while n.abs() >= 365 {
+            // Find the number of years in n days.
+            let k = (n / 365) as i32;
+            if year > 0 && year + k < 0 {
+                n += 1;
+            }
+            year += k;
+
+            if k > 0 {
+                n -= (k.abs() * 365 + Self::number_leap_years(year - 1)
+                    - Self::number_leap_years(year - k)) as i64;
+            } else {
+                n += (k.abs() * 365 + Self::number_leap_years(year - k)
+                    - Self::number_leap_years(year + 1)) as i64;
+            }
+        }
+
+        if n < 0 {
+            // If the remaining number of days is negative, change the year and count the complement.
+            year -= 1;
+            n += if Self::is_leap_year(year) { 366 } else { 365 };
+        }
+
+        if Self::is_leap_year(year) && n >= 60 {
+            n -= 1;
+        } else if Self::is_leap_year(year) && n > 31 {
+            return ((n - 30) as i8, 2, year);
+        }
+
+        let q = DAY_IN_MONTH // Find month.
+            .iter()
+            .enumerate()
+            .filter(|&(_, &x)| i64::from(x) <= n)
+            .max_by_key(|&(_, &x)| x)
+            .map(|(index, &_value)| index + 1)
+            .unwrap();
+
+        ((n - DAY_IN_MONTH[q - 1] as i64) as i8 + 1, q as i8, year)
+    }
+
+    /// Checks whether the year is leap year.
+    fn is_leap_year(year: i32) -> bool {
+        (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0)
+    }
+
+    #[napi(js_name = "toString")]
+    pub fn to_format(&self) -> String {
+        self.to_string()
+    }
+
+    /// Returns the number of days in a month.
+    fn days_in_month(month: i8, year: i32) -> i8 {
+        match month {
+            1 | 3 | 5 | 7 | 8 | 10 | 12 => 31,
+            4 | 6 | 9 | 11 => 30,
+            2 => {
+                if Self::is_leap_year(year) {
+                    29
+                } else {
+                    28
+                }
+            }
+            _ => 0,
+        }
+    }
+
+    fn number_leap_years(n: i32) -> i32 {
+        n / 4 - n / 100 + n / 400
+    }
```

Comment:
In each case we need a number of leap years between some range. Maybe it would be cleaner to have a function that does that rather than calling it twice?

Reply:
in `days_since_epoch` this function is called once.